src/realm/parser/parser.cpp

```diff
@@ -0,0 +1,491 @@
+////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2015 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+#include "parser.hpp"
+
+#include <iostream>
+
+#include <pegtl.hpp>
+#include <pegtl/analyze.hpp>
+#include <pegtl/contrib/tracer.hpp>
+
+// String tokens can't be followed by [A-z0-9_].
+#define string_token_t(s) seq< TAOCPP_PEGTL_ISTRING(s), not_at< identifier_other > >
+
+using namespace tao::pegtl;
+
+namespace realm {
+namespace parser {
+
+// strings
+struct unicode : list< seq< one< 'u' >, rep< 4, must< xdigit > > >, one< '\\' > > {};
+struct escaped_char : one< '"', '\'', '\\', '/', 'b', 'f', 'n', 'r', 't', '0' > {};
+struct escaped : sor< escaped_char, unicode > {};
+struct unescaped : utf8::range< 0x20, 0x10FFFF > {};
+struct chars : if_then_else< one< '\\' >, must< escaped >, unescaped > {};
+struct dq_string_content : until< at< one< '"' > >, must< chars > > {};
+struct dq_string : seq< one< '"' >, must< dq_string_content >, any > {};
+
+struct sq_string_content : until< at< one< '\'' > >, must< chars > > {};
+struct sq_string : seq< one< '\'' >, must< sq_string_content >, any > {};
+
+// base64 encoded data
+struct b64_allowed : sor< disable< alnum >, one< '/' >, one< '+' >, one< '=' > > {};
+struct b64_content : until< at< one< '"' > >, must< b64_allowed > > {};
+struct base64 : seq< TAOCPP_PEGTL_ISTRING("B64\""), must< b64_content >, any > {};
+
+// numbers
+struct minus : opt< one< '-' > > {};
+struct dot : one< '.' > {};
+
+struct float_num : sor<
+    seq< plus< digit >, dot, star< digit > >,
+    seq< star< digit >, dot, plus< digit > >
+> {};
+struct hex_num : seq< one< '0' >, one< 'x', 'X' >, plus< xdigit > > {};
+struct int_num : plus< digit > {};
+
+struct number : seq< minus, sor< float_num, hex_num, int_num > > {};
+
+struct timestamp_number : disable< number > {};
+struct first_timestamp_number : disable< timestamp_number > {};
```

Comment:
What's the difference between these two rules?

Reply:
There is no syntactic difference between those rules, the difference is in how the matches are processed (the action methods later on in this file). The rules for timestamps build up a list of the number inputs in a vector of strings to be converted to actual integers later on. Whenever a `timestamp_number` is matched, it is added to the state and when the full `timestamp` rule is matched, this stack is processed  to build an expression containing a `Timestamp` value. The problem with this is that a `timestamp_number` match can be triggered without actually matching a complete `timestamp` rule, for example on a column named "T2015". The `first_timestamp_number` is there to tell the state to clear that list and place the new match on the stack. I'm not yet familiar enough with the PEGTL rules to know if there is a more efficient way to achieve this.